File: backend/src/main/java/com/hgw/gestionale/model/controller/ModelController.java
package com.hgw.gestionale.model.controller;

import com.hgw.gestionale.model.dto.CreateModelRequest;
import com.hgw.gestionale.model.dto.ModelResponse;
import com.hgw.gestionale.model.dto.UpdateModelRequest;
import com.hgw.gestionale.model.service.ModelService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping(value = "/api/models", produces = MediaType.APPLICATION_JSON_VALUE)
@RequiredArgsConstructor
@Slf4j
public class ModelController {
    private final ModelService modelService;

    @PostMapping
    public ModelResponse createModel(@Valid @RequestBody CreateModelRequest request) {
        log.info("ModelController.create creazione model: {}", request.nome());
        return modelService.create(request);
    }

    @GetMapping
    public List<ModelResponse> getAllModels() {
        log.info("ModelController.getAll richiesta lista model");
        return modelService.getAll();
    }

    @GetMapping("/{id}")
    public ModelResponse getModelById(@PathVariable Long id) {
        log.info("ModelController.getModelById richiesta model id={}", id);
        return modelService.getById(id);
    }

    @GetMapping("/by-brand-id/{id}")
    public List<ModelResponse> getModelsByBrandId(@PathVariable Long id) {
        log.info("ModelController.getModelByBrandId richiesta model id={}", id);
        return modelService.getByBrandId(id);
    }

    @GetMapping("/by-name/{nome}")
    public ModelResponse getModelByName(@PathVariable String nome) {
        log.info("ModelController.getByName richiesta model nome={}", nome);
        return modelService.getByName(nome);
    }

    @GetMapping("/search")
    public List<ModelResponse> searchModel(@RequestParam String nome) {
        log.info("ModelController.searchModel ricerca model nome={}", nome);
        return modelService.search(nome);
    }

    @GetMapping("/search/by-brand")
    public List<ModelResponse> searchModelByBrand(
            @RequestParam Long brandId,
            @RequestParam String nome
    ) {
        log.info("ModelController.searchModelByBrand ricerca model by brand nome={} brand={}", nome, brandId);
        return modelService.searchByBrand(brandId, nome);
    }

    @PutMapping("/{id}")
    public ModelResponse updateModel(
            @PathVariable Long id,
            @Valid @RequestBody UpdateModelRequest request
    ) {
        log.info("ModelController.update aggiornamento model id={}", id);
        return modelService.update(id, request);
    }

    @DeleteMapping("/{id}")
    public void deleteModel(@PathVariable Long id) {
        log.info("ModelController.delete eliminazione model id={}", id);
        modelService.delete(id);
    }
}

File: backend-node/src/controllers/modelController.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  OperationId,
  Path,
  Post,
  Put,
  Query,
  Route,
  Tags,
} from "tsoa";
import { ModelResponse, CreateModelRequest, UpdateModelRequest } from "../dto/model.dto";

@Route("models")
@Tags("Models")
export class ModelController extends Controller {
  @Post("/")
  @OperationId("createModel")
  public async createModel(@Body() body: CreateModelRequest): Promise<ModelResponse> {
    throw new Error("Not implemented");
  }

  @Get("/")
  @OperationId("getAllModels")
  public async getAllModels(): Promise<ModelResponse[]> {
    throw new Error("Not implemented");
  }

  @Get("/search")
  @OperationId("searchModels")
  public async searchModels(@Query() nome: string): Promise<ModelResponse[]> {
    throw new Error("Not implemented");
  }

  @Get("/search/by-brand")
  @OperationId("searchModelsByBrand")
  public async searchModelsByBrand(
    @Query() brandId: number,
    @Query() nome: string
  ): Promise<ModelResponse[]> {
    throw new Error("Not implemented");
  }

  @Get("/search/by-brand-name")
  @OperationId("searchModelsByBrandName")
  public async searchModelsByBrandName(
    @Query() brandNome: string,
    @Query() nome: string
  ): Promise<ModelResponse[]> {
    throw new Error("Not implemented");
  }

  @Get("/by-brand-id/{id}")
  @OperationId("getModelsByBrandId")
  public async getModelsByBrandId(@Path() id: number): Promise<ModelResponse[]> {
    throw new Error("Not implemented");
  }

  @Get("/by-name/{nome}")
  @OperationId("getModelByName")
  public async getModelByName(@Path() nome: string): Promise<ModelResponse> {
    throw new Error("Not implemented");
  }

  @Get("/{id}")
  @OperationId("getModelById")
  public async getModelById(@Path() id: number): Promise<ModelResponse> {
    throw new Error("Not implemented");
  }

  @Put("/{id}")
  @OperationId("updateModel")
  public async updateModel(
    @Path() id: number,
    @Body() body: UpdateModelRequest
  ): Promise<ModelResponse> {
    throw new Error("Not implemented");
  }

  @Delete("/{id}")
  @OperationId("deleteModel")
  public async deleteModel(@Path() id: number): Promise<void> {
    throw new Error("Not implemented");
  }
}
